src/lmdb.mli

```diff
@@ -173,27 +174,31 @@ end
 module type S = sig
 
   (** A handle for an individual database. *)
-  type t
+  type -'cap t
 
   (** The key of the database. *)
   type key
 
   (** The values of the database. *)
   type elt
 
+  type 'cap create_mode
+  val create_db : [> `Read | `Write ] create_mode
+  val open_db : [> `Read ] create_mode
+
   (** [create env "foo"] open the database ["foo"] in the environment [env].
 
       @param create if [true], the database will be created if it doesn't
       exists. Invalid if [env] is read only.
 
       @raise Not_found if the database doesn't exist. and [create] wasn't [true].
   *)
-  val create : ?create:bool -> Env.t -> string -> t
+  val create : 'cap create_mode -> ?txn:'cap Txn.t -> 'cap Env.t -> string -> 'cap t
```

Comment:
The open/create flag has no influence on the return type of the ``create`` function, but it will require that the environment and transaction are read/write when trying to create a dbi.

I think we can expose the GADT type, this will allow to use ``Open`` / ``Create`` flags. I don't like the idea to duplicate the function.

``env`` and ``db`` values are returned monomorphic.

Reply:
The signature is `'cap create_mode -> ?txn:'cap Txn.t -> 'cap Env.t -> string -> 'cap t`, how can you say that `'cap create_mod` has no influence of `'cap t` ?!

Furthermore, the type coming from create_mod is either ``[> `Read]`` or ``[> `Read | `Write]``. Those types are *not* monomorphic. There is a row type variable hidden in the `>`, which you can clearly see if you try in the toplevel to open an environment (it will give you a weak type).

Finally, I insist, If I have to choose between 1 GADT + 1 function and 2 functions, I'll pick the later. Not everyone is GADT-fluent, and the API must be accessible to them as well.

Reply:
In ``'cap Env.t -> 'cap t`` ``'cap Env.t`` is monomorphic, therefore ``'cap t`` will be monomorphic, too.
I'm afraid you saw weak types in the toplevel, because of a copy-paste glitch I introduced in commit f95b972. Please excuse this error on my part:

``` ocaml
  val create : 'cap create_mode ->
    conv_key: 'k Conv.t ->
    conv_val: 'v Conv.t ->
    ?flags:   Flags.t ->
    ?txn:     ([< `Read | `Write ] as 'p) Txn.t ->
    ?name:    string ->
    'p Env.t ->
    ('k, 'v, 'cap) t
```
This should have been ``'cap Env.t`` instead of ``'p Env.t``. Same for ``'cap Txn.t``.

Now you'll only get monomorphic types in the toplevel:

``` ocaml
# open Lmdb;;
# let env = Env.create rw "/tmp/";;
val env : [ `Read | `Write ] Lmdb.Env.t = <abstr>
# Db.create;;
- : 'cap Lmdb.Db.create_mode ->
    ?txn:'cap Lmdb.Txn.t -> ?name:string -> 'cap Lmdb.Env.t -> 'cap Lmdb.Db.t
= <fun>
# Db.(create create_db);;
- : ?txn:(_[> `Read | `Write ] as 'a) Lmdb.Txn.t ->
    ?name:string -> 'a Lmdb.Env.t -> 'a Lmdb.Db.t
= <fun>
# Db.(create open_db);;
- : ?txn:(_[> `Read ] as 'a) Lmdb.Txn.t ->
    ?name:string -> 'a Lmdb.Env.t -> 'a Lmdb.Db.t
= <fun>
# Db.(create open_db env);;
- : [ `Read | `Write ] Lmdb.Db.t = <abstr>
# Db.(create create_db env);;
- : [ `Read | `Write ] Lmdb.Db.t = <abstr>
```
Please also note how the ``'cap create_mode`` parameter, due to its _open_ variant type enforces a lower bound on the capabilities of ``txn`` resp. ``env``, but has no further influence on the return type.

> Not everyone is GADT-fluent, and the API must be accessible to them as well.

I'm certainly no GADT guru either and  at the moment no GADT is exposed in the interface. In the implementation they could easily be replace by ``bool`` or ``int``. It just might be nicer to offer ``Constructors`` instead of ``values`` to the user. But this is bikeshedding…

Reply:
Right, so it indeed works, but it is really far too clever in my opinion. Please split the function in two (the underlying implementation can still be shared) and make the types explicit for those two, which will solve the problem, ensure that it doesn't break and simplify the API.

Reply:
I put in two additional create functions you request, but kept my original one, so you can compare side-by-side. A parameter requesting to create on demand is idiomatic and used widely. The typing magic is nothing the user will have to worry about. When we change the names it even reads nicely:
``Db.create new_db env`` or
``Db.create existing_db env``
If we are honest, the whole capabilities phantom typing is overengineered, but it _is_ clever ;-)

Reply:
> If we are honest, the whole capabilities phantom typing is overengineered, but it _is_ clever ;-)

I'm well aware, that's why I'm trying to reign in a bit your enthusiasm about typing cleverness. ;)